File: src/com/fern/110-Balanced-Binary-Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Time complexity: O(n);
// space complexity: O(n)

// Method: DFS Postorder

// Notes: Because it will get complicated if we try to return a booleon and a int from our
//        functions. It is best to return the current height and "-1" which represents false.

class Solution {
    public boolean isBalanced(TreeNode root) {
        if(root == null) return true;
        return depth(root) != -1;

    }

    public int depth(TreeNode node){
        if(node == null) return 0;

        int left = depth(node.left);
        if(left == -1) return -1;

        int right = depth(node.right);
        if(right == -1) return -1;

        if(Math.abs(left - right) > 1) return -1;
        return Math.max(left,right) + 1;

    }
}

File: src/JavaScript-Solutions/110-Balanced-Binary-Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

var isBalanced = function(root) {
    
 function branchHeight(root){
     if(!root) return 0;
     
     let left = branchHeight(root.left);
     let right = branchHeight(root.right);
     
     if(left === -1 || 
        right === -1 || 
        Math.abs(right - left) > 1) return -1;  
     
     return Math.max(left, right) + 1;
 }
 
 
 return branchHeight(root) !== -1 
};